cdc/verification/checker.go

```diff
@@ -0,0 +1,218 @@
+	rows, err := c.db.QueryContext(ctx, "SHOW DATABASES")
+	if err != nil {
+		return nil, cerror.WrapError(cerror.ErrMySQLQueryError, err)
+	}
+	defer func() {
+		if err = rows.Close(); err != nil {
+			log.Error("getAllDBs close rows failed", zap.Error(err))
+		}
+	}()
+
+	dbs := []string{}
+	for rows.Next() {
+		var d string
+		if err = rows.Scan(&d); err != nil {
+			return dbs, cerror.WrapError(cerror.ErrMySQLQueryError, err)
+		}
+
+		dbs = append(dbs, d)
+	}
+
+	return dbs, nil
+}
+
+func (c *checker) getAllTables(ctx context.Context, db string, f *filter.Filter) ([]string, error) {
+	rows, err := c.db.QueryContext(ctx, "SHOW TABLES")
+	if err != nil {
+		return nil, cerror.WrapError(cerror.ErrMySQLQueryError, err)
+	}
+	defer func() {
+		if err = rows.Close(); err != nil {
+			log.Error("getAllTables close rows failed", zap.Error(err))
+		}
+	}()
+
+	tables := []string{}
+	for rows.Next() {
+		var t string
+		if err = rows.Scan(&t); err != nil {
+			return tables, cerror.WrapError(cerror.ErrMySQLQueryError, err)
+		}
+		if f.ShouldIgnoreTable(db, t) {
+			continue
+		}
+		tables = append(tables, t)
+	}
+
+	return tables, nil
+}
+
+type columnInfo struct {
+	Field   string
+	Type    string
+	Null    string
+	Key     string
+	Default *string
+	Extra   string
+}
+
+func (c *checker) getColumns(ctx context.Context, tableName string) ([]columnInfo, error) {
+	rows, err := c.db.QueryContext(ctx, fmt.Sprintf("SHOW COLUMNS FROM %s", tableName))
+	if err != nil {
+		return nil, cerror.WrapError(cerror.ErrMySQLQueryError, err)
+	}
+	defer func() {
+		if err = rows.Close(); err != nil {
+			log.Error("getColumns close rows failed", zap.Error(err))
+		}
+	}()
+
+	result := []columnInfo{}
+	for rows.Next() {
+		var t columnInfo
+		if err = rows.Scan(&t.Field, &t.Type, &t.Null, &t.Key, &t.Default, &t.Extra); err != nil {
+			return result, cerror.WrapError(cerror.ErrMySQLQueryError, err)
+		}
+		result = append(result, t)
+	}
+
+	return result, nil
+}
+
+func (c *checker) doChecksum(ctx context.Context, columns []columnInfo, databaseName, tableName string) (string, error) {
+	var columnNames, isNull []string
+	for _, item := range columns {
+		columnNames = append(columnNames, item.Field)
+
+		t := fmt.Sprintf("ISNULL(%s)", item.Field)
+		isNull = append(isNull, t)
+	}
+
+	a := strings.Join(columnNames, ",")
+	b := strings.Join(isNull, ",")
+
+	// ref: https://www.percona.com/doc/percona-toolkit/LATEST/pt-table-checksum.html
+	// TODO: hash function as a option
+	concat := fmt.Sprintf("CONCAT_WS(',', %s, %s)", a, b)
+	query := fmt.Sprintf("SELECT BIT_XOR(CAST(crc32(%s) AS UNSIGNED)) AS checksum FROM %s", concat, tableName)
+
+	log.Debug("do checkSum",
+		zap.String("db", databaseName),
+		zap.String("table", tableName),
+		zap.String("query", query))
+	var checkSum string
+	err := c.db.QueryRowContext(ctx, query).Scan(&checkSum)
+	return checkSum, cerror.WrapError(cerror.ErrMySQLQueryError, err)
+}
+
+// TODO: use ADMIN CHECKSUM TABLE for tidb if needed
```

Comment:
Result of `admin checksum table` contains row keys, so they can be different even if both upstream and downstream are TiDB. I think we can remove this comment.

Reply:
is it safe if we just compare Total_kvs, Total_kvs, skip  Checksum_crc64_xor
```
mysql> ADMIN CHECKSUM TABLE t1;
+---------+------------+----------------------+-----------+-------------+
| Db_name | Table_name | Checksum_crc64_xor   | Total_kvs | Total_kvs |
+---------+------------+----------------------+-----------+-------------+
| test    | t1         | 10909174369497628533 |         3 |          75 |
```